GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -790,13 +798,46 @@ private void ClearProjectionCaches()
 
         private void RefreshSparseFolders()
         {
+            this.sparseFoldersNeedingRefresh = null;
             this.rootSparseFolder.Children.Clear();
             if (this.sparseCollection != null)
             {
+                string pathSeparatorString = Path.DirectorySeparatorChar.ToString();
+
+                this.sparseFoldersNeedingRefresh = new HashSet<string>(GVFSPlatform.Instance.Constants.PathComparer);
+
+                // To avoid a lot of no-op adds, always refresh the root
+                this.sparseFoldersNeedingRefresh.Add(string.Empty);
+
                 Dictionary<string, SparseFolderData> parentFolder = this.rootSparseFolder.Children;
-                foreach (string directoryPath in this.sparseCollection.GetAll())
+                HashSet<string> oldSparsePaths = this.sparsePaths;
+                this.sparsePaths = this.sparseCollection.GetAll();
+                foreach (string directoryPath in this.sparsePaths)
                 {
                     string[] folders = directoryPath.Split(new[] { Path.DirectorySeparatorChar }, StringSplitOptions.RemoveEmptyEntries);

```

Comment:
Why are we using `Split` here and then `LastIndexOf` below?  Are there differences in memory/performance?  Can we use a common method for both?

Reply:
Good point, these started out as different because in my first iteration I was only grabbing the immediate parent.  Now that we're processing all ancestors we should be able to make this code consistent.